UMN-VMWareRA.psm1

```diff
@@ -1197,6 +1197,321 @@ function Set-VMWareRAVMTagAttachedToObject {
+
+    .PARAMETER cluster
+        Name of cluster to return list of VMs from (able to return more then 1000 VMs), case sensitivity required.
+
+    .PARAMETER clusterID
+        ID of cluster to return list of VMs from (able to return more then 1000 VMs.) Overrides cluster parameter if specified.
+
+    .OUTPUTS
+        List of VM objects based on parameters specified.
+
+    .Notes
+        Author: Aaron Smith
+#>
+    [CmdletBinding()]
+    Param
+    (
+        [Parameter(Mandatory)]
+        [string]$vCenter,
+        
+        [Parameter(Mandatory)]
+        [string]$sessionID,
+
+        [string]$name,
+
+        [string]$host,
+
+        [string]$hostID,
+
+        [string]$cluster,
+
+        [string]$clusterID
+    )
+
+    Begin
+    {
+    }
+    Process
+    {
+        if ($cluster)
+        {
+            $clusterID = (Get-VMWareRACluster -vCenter $vCenter -sessionID $sessionID -name $cluster).cluster;
+
+            if (!$clusterID)
+            {
+                throw "Cluster ${cluster} not found by name";
+            }
+        }
+
+        if ($host)
+        {
+            $hostID = (Get-VMWareRAHost -vCenter $vCenter -sessionID $sessionID -name $host).host;
+
+            if (!$hostID)
+            {
+                throw "VMHost ${host} not found by name";
+            }
+        }
+
+        ## List of VMs compiled from queries for return
+        [Array] $vmList = @();
+
+        ## Construct url
+        $url = "https://$vCenter/rest/vcenter/vm"
+
+        [Array] $urlFilters = @();
+
+        if ($hostID)
+        {
+            $urlFilters += "filter.hosts=${hostID}";
+        }
+
+        if ($urlFilters.Count -gt 0)
+        {
+            $url += "?";
+
+            foreach ($urlFilterItem in $urlFilters)
+            {
+                $url += $urlFilterItem + "&";
+            }
+
+            $return = Invoke-WebRequest -Uri $url -Method Get -Headers @{'vmware-api-session-id'=$sessionID} -ContentType 'application/json' -UseBasicParsing
+            $vmList += (($return.Content | ConvertFrom-Json).value)
+        }
+        else
+        {
+            [Array] $hostList = @();
+
+            if ($clusterID)
+            {
+                $hostList += Get-VMWareRAHost -vCenter $vCenter -sessionID $sessionID -clusterID $clusterID;
+            }
+            else
+            {
+                $hostList += Get-VMWareRAHost -vCenter $vCenter -sessionID $sessionID;
+            }
+
+            foreach ($hostItem in $hostList)
+            {
+                $hostID = $hostItem.host;
+                $url = "https://$vCenter/rest/vcenter/vm?filter.hosts=${hostID}"
+
+                $return = Invoke-WebRequest -Uri $url -Method Get -Headers @{'vmware-api-session-id'=$sessionID} -ContentType 'application/json' -UseBasicParsing
+                $vmList += (($return.Content | ConvertFrom-Json).value);
+            }
+        }
+
+        if ($name)
+        {
+            [Array] $filteredVMList = @();
```

Comment:
Arrays in powershell are fine if they are static, but horrible if you are going to add a lot of things one by one, use an [System.Collections.ArrayList]   and when you add, use $myArra.add($thing)

Reply:
Good to know, thanks!  If you prefer I can modify and submit a new pull request, or if everything looks good you can merge then immediately make code-formatting changes and switch use of [Array] to [ArrayList].